src/Parallel/LockfreeUnorderedSet.hpp

```diff
@@ -0,0 +1,451 @@
+      }
+    }
+
+    // We now know that _if_ the element exists, it's in what was in the first
+    // empty slot, which means we either encounter the element in the first
+    // available slot or we get to insert it in that slot.
+    for (size_t i = this_bucket_size.load(std::memory_order_relaxed);
+         i < (bucket_size - 1); ++i) {
+      const size_t index = bucket_index + 1 + i;
+      const std::uint64_t probed_internal_key =
+          entries_[index].load(std::memory_order_relaxed);
+      if (probed_internal_key == internal_key) {
+        return true;  // set already contains value
+      } else {
+        if (probed_internal_key != EmptySlotValue) {
+          // The entry is used by another key.
+          continue;
+        }
+        // The entry is empty. Let's try to set it.
+        std::uint64_t current_key_in_slot = EmptySlotValue;
+        if (not entries_[index].compare_exchange_strong(
+                current_key_in_slot, internal_key, std::memory_order_relaxed,
+                std::memory_order_relaxed) and
+            current_key_in_slot != internal_key) {
+          // Another thread just stole this slot from us. Try next slot.
+          continue;
+        }
+        size_.fetch_add(1, std::memory_order_relaxed);
+        if (this_bucket_size.load(std::memory_order_relaxed) <
+            bucket_size - 1) {
+          this_bucket_size.fetch_add(1, std::memory_order_release);
+        }
+        // Successful insert. Return.
+        return true;
+      }
+    }
+  } else {
+    // We first do a linear search in the bucket to see if the element exists.
+    for (size_t i = 0; i < bucket_size; ++i) {
+      const size_t index = bucket_index + i;
+      const std::uint64_t probed_internal_key =
+          entries_[index].load(std::memory_order_relaxed);
+      if (probed_internal_key == internal_key) {
+        return true;  // set already contains value
+      }
+    }
+
+    // We now know that _if_ the element exists, it's in what was in the first
+    // empty slot, which means we either encounter the element in the first
+    // available slot or we get to insert it in that slot.
+    for (size_t i = 0; i < bucket_size; ++i) {
+      const size_t index = bucket_index + i;
+      const std::uint64_t probed_internal_key =
+          entries_[index].load(std::memory_order_relaxed);
+      if (probed_internal_key != EmptySlotValue) {
+        // The entry is used by another key.
+        continue;
+      }
+      // The entry is empty. Let's try to set it.
+      std::uint64_t current_key_in_slot = EmptySlotValue;
+      if (not entries_[index].compare_exchange_strong(
+              current_key_in_slot, internal_key, std::memory_order_relaxed,
+              std::memory_order_relaxed) and
+          current_key_in_slot != internal_key) {
+        // Another thread just stole this slot from us. Try next slot.
+        continue;
+      }
+      size_.fetch_add(1, std::memory_order_relaxed);
```

Comment:
If the compare-exchange failed but `current_key_in_slot == internal_key`, you should skip the modification of `size_`.  Also in the TrackBucketSize version.